spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/metrics/GcpStackdriverMetricsAutoConfiguration.java

```diff
@@ -0,0 +1,74 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.autoconfigure.metrics;
+
+import java.io.IOException;
+
+import com.google.api.gax.core.CredentialsProvider;
+import io.micrometer.core.instrument.Clock;
+import io.micrometer.core.instrument.step.StepMeterRegistry;
+import io.micrometer.stackdriver.StackdriverConfig;
+
+import org.springframework.boot.actuate.autoconfigure.metrics.export.stackdriver.StackdriverMetricsExportAutoConfiguration;
+import org.springframework.boot.actuate.autoconfigure.metrics.export.stackdriver.StackdriverProperties;
+import org.springframework.boot.autoconfigure.AutoConfigureBefore;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.cloud.gcp.core.DefaultCredentialsProvider;
+import org.springframework.cloud.gcp.core.GcpProjectIdProvider;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+/**
+ * Provides auto-detection for project.
+ *
+ * @author Eddú Meléndez
+ */
+@Configuration
+@AutoConfigureBefore(StackdriverMetricsExportAutoConfiguration.class)
+@ConditionalOnClass({StepMeterRegistry.class, StackdriverConfig.class})
+@ConditionalOnBean(Clock.class)
+@EnableConfigurationProperties({GcpMetricsProperties.class, StackdriverProperties.class})
+@ConditionalOnProperty(value = "spring.cloud.gcp.metrics.enabled", matchIfMissing = true)
+public class GcpStackdriverMetricsAutoConfiguration {
+
+	private final StackdriverProperties stackdriverProperties;
+
+	private final String projectId;
+
+	private final CredentialsProvider credentialsProvider;
+
+	public GcpStackdriverMetricsAutoConfiguration(GcpMetricsProperties gcpMetricsProperties,
+			StackdriverProperties stackdriverProperties, GcpProjectIdProvider gcpProjectIdProvider,
+			CredentialsProvider credentialsProvider) throws IOException {
+		this.stackdriverProperties = stackdriverProperties;
+		this.projectId = (gcpMetricsProperties.getProjectId() != null)
+				? gcpMetricsProperties.getProjectId() : gcpProjectIdProvider.getProjectId();
+		this.credentialsProvider = gcpMetricsProperties.getCredentials().hasKey()
+				? new DefaultCredentialsProvider(gcpMetricsProperties) : credentialsProvider;
+	}
+
+	@Bean
+	@ConditionalOnMissingBean
+	public StackdriverConfig stackdriverConfig() {
```

Comment:
do we also need to produce the `StackdriverMeterRegistry`? Or has Spring Boot Actuator been updated to produce it automatically? (It was an open issue)

Reply:
it is not required since we are performing this auto-configuration before the one provided by spring-boot. So, we just need to pre-configure the config so spring-boot takes it

Reply:
just added an assertion to make sure the `meterRegistry` bean is created

Reply:
last time i used this, i needed to create my own registry bean, and documented here: https://micrometer.io/docs/registry/stackdriver

https://github.com/micrometer-metrics/micrometer-docs/pull/115#pullrequestreview-380850792

I'm unsure if actuator has been updated to autoconfigure the registry bean

Reply:
this is the current auto-configuration for stackdriver in spring-booot https://github.com/spring-projects/spring-boot/blob/master/spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/stackdriver/StackdriverMetricsExportAutoConfiguration.java

Reply:
Oh nice!! Fixed in 2.3 👏👏